src/HttpOverStream.Server.Owin/CustomListenerHost.cs

```diff
@@ -179,6 +188,99 @@ public void Initialize(IAppBuilder builder)
             }
         }
 
-    }
+        private class WriteInterceptStream : Stream
+        {
+            private readonly Stream _innerStream;
+            private Func<Task> _once;
+
+            public WriteInterceptStream(Stream innerStream, Func<Task> once)
+            {
+                _innerStream = innerStream;
+                _once = once;
+            }
+
+            public override void Flush()
+            {
+                var once = Interlocked.Exchange(ref _once, null);
+                if (once != null)
+                {
+                    once().Wait();

```

Comment:
Should we be using synchronous waits?

Reply:
On synchronous writes/flushes, yes (but it should never happen with Asp.Net Web API middleware)